src/solid-logic-move-me.ts

```diff
@@ -0,0 +1,181 @@
+    await this.store.fetcher
+      .load(profileDocument, { withCredentials: false, cache: 'reload' })
+  }
+
+  async loadPreferences (me: NamedNode): Promise<NamedNode> {
+    const preferencesFile = this.store.any(me, ns.space('preferencesFile'))
+
+    /**
+     * Are we working cross-origin?
+     * Returns True if we are in a webapp at an origin, and the file origin is different
+     */
+    function differentOrigin (): boolean {
+      return `${window.location.origin}/` !== preferencesFile.site().uri
+    }
+
+    if (!preferencesFile) {
+      throw new Error(`Can't find a preference file pointer in profile ${me.doc()}`)
+    }
+
+    // //// Load preference file
+    try {
+      this.store.fetcher
+        .load(preferencesFile, { withCredentials: true })
+    } catch (err) {
+      // Really important to look at why
+      const status = err.status
+      debug.log(
+        `HTTP status ${status} for preference file ${preferencesFile}`
+      )
+      if (status === 401) {
+        throw new UnauthorizedError()
+      }
+      if (status === 403) {
+        if (differentOrigin()) {
+          throw new CrossOriginForbiddenError()
+        }
+        throw new SameOriginForbiddenError()
+      }
+      if (status === 404) {
+        throw new NotFoundError(preferencesFile)
+      }
+      throw new FetchError(err.status, err.message)
+    }
+    return preferencesFile
+  }
+
+  getTypeIndex (me: NamedNode | string, preferencesFile: NamedNode | string, isPublic: boolean): NamedNode[] {
+    return this.store.each(
+      me,
+      (isPublic ? ns.solid('publicTypeIndex') : ns.solid('privateTypeIndex')),
+      undefined,
+      preferencesFile
+    )
+  }
+
+  getContainerElements (cont: NamedNode) {
+    return this.store.each(cont, ns.ldp('contains'))
+  }
+
+  getRegistrations (instance, theClass) {
+    return this.store
+      .each(undefined, ns.solid('instance'), instance)
+      .filter((r) => {
+        return this.store.holds(r, ns.solid('forClass'), theClass)
+      })
+  }
+
+  load (doc: NamedNode | string) {
+    return this.store.fetcher.load(doc)
+  }
+
+  async createEmptyRdfDoc (doc: NamedNode, comment: string) {
+    await this.store.fetcher.webOperation('PUT', doc.uri, {
+      data: `# ${new Date()} ${comment}
+`,
+      contentType: 'text/turtle'
+    })
+  }
+
+  // @@@@ use the one in rdflib.js when it is available and delete this
+  updatePromise (
+    del: Array<Statement>,
+    ins: Array<Statement> = []
+  ): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.store.updater.update(del, ins, function (uri, ok, errorBody) {
+        if (!ok) {
+          reject(new Error(errorBody))
+        } else {
+          resolve()
+        }
+      }) // callback
+    }) // promise
+  }
+
+  clearStore () {
+    this.store.statements.slice().forEach(this.store.remove.bind(this.store))
+  }
+}
+
+export class UnauthorizedError extends Error {
```

Comment:
I can't say much about the PR in general because I don't know the codebase, but I spotted a potential improvement here :).

I think it's better to define a custom base error (for example SolidLogicError) and extend everything from that. That way, it's not necessary to override the constructor in each error subclass.

Check out these files in soukai: [SoukaiError.ts](https://github.com/NoelDeMartin/soukai/blob/7f2575e20cbec8ddc9041da9436c2f69fa2b8696/src/errors/SoukaiError.ts) and [SoukaiError.test.ts](https://github.com/NoelDeMartin/soukai/blob/7f2575e20cbec8ddc9041da9436c2f69fa2b8696/src/errors/SoukaiError.test.ts).

Reply:
Nice! Thanks